pysmt/oracles.py

```diff
@@ -30,6 +30,73 @@
 from pysmt.logics import Logic, Theory, get_closer_pysmt_logic
 
 
+class SizeOracle(walkers.DagWalker):
+    """Evaluates the size of a formula"""
+
+    # Counting type can be:
+    # - TREE_NODES : counts the number of nodes in the formula seen as a tree
+    # - DAG_NODES  : counts the number of nodes in the formula seen as a DAG
+    # - LEAVES     : counts the number of leaves in the formula seen as a tree
+    # - DEPTH      : counts the maximum number of levels in the formula
+    (COUNTING_TYPE_TREE_NODES,
+     COUNTING_TYPE_DAG_NODES,
+     COUNTING_TYPE_LEAVES,
+     COUNTING_TYPE_DEPTH) = range(4)
+
+    def __init__(self, env=None):
+        walkers.DagWalker.__init__(self, env=env)
+
+        # Clear the mapping function
+        self.functions.clear()
+
+        # Propagate truth value, and force False when a Quantifier
+        # is found.
+        for elem in op.ALL_TYPES:
+            self.functions[elem] = self.walk_count
+
+        # Check that no operator in undefined
+        assert self.is_complete(verbose=True)
+
+
+    def _get_key(self, formula, counting_type):
+        # Memoize using a tuple (counting_type, formula)
+        return (counting_type, formula)
+
+
+    def get_size(self, formula, counting_type=None):
```

Comment:
Each time this function is called, we only use one of the possible "counting_types"/metric. Therefore, it seems inconvenient to have to check which counting_type we are using within walk_count for every node.
I'd suggest that this method sets the value of walk_count.

```
def get_size(...):
  self.set_walk_method(counting_type)
  res = self.walk(formula)
  self.walk_count = self.walk_error
```

where self.set_walk_method assigns the appropriate function to all node types (as done during init())